rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+was likely severed and a number of operations timed out before the connection
+could be re-established.
+
+Determining why the connection was severed requires correlation to other
+information.  For example, we may need to look at the Couchbase Server logs to
+see that a particular process for a given service crashed where we would expect
+to find the SDK runtime/instance identifier being sent in the request (defined
+in the [Response Time Observability RFC](rfc/0035-rto.md)).
+
+We should note that existence of timeouts is possibly expected.  For example,
+with an ad-targeting or recommendation targeting deployment, there may be short
+latencies that when not serviced, lead to other processing.  This is worth
+mentioning because it will be important to compare a deployment to itself, and
+there may be limited value in comparing across deployments.
+
+## Capella Service Level Metrics
+
+The counters described above give a good picture of actual impact on a deployed
+app, but they do not give a sense of whether or not a given deployment is
+meeting an expected service level.  Said another way, a Couchbase Capella
+Cluster deployment could be operational, but very slow during a rebalance
+leading to user discontent, but never turn into an error.
+
+Service levels are specific to the user's targets.  While it's impossible to
+properly measure whether or not an end-user's app is meeting service level
+targets, we can establish a predefined histogram that would allow Couchbase to
+simply and inexpensively understand if the service level has likely changed.
+
+It's proposed that we establish a set of standardized histograms, specific to
+services and to the service level expected with the service using those
+parameters.  Counters for these histograms can be sent to Couchbase and both
+stored individually and aggregated.
+
+As an aggregated set of metrics, it would give Couchbase a rough view over the
+service level the user is experiencing.  Deviations in service level can be dug
+into and, in particular, we can check for correlation to Couchbase Capella
+management activities like upgrades.
+
+This list represents the services, buckets.
+
+* Data Service, Non-Durable
+  * Retrieval
+    * <1ms
+    * 1-10ms
+    * 10-100ms
+    * 100-500ms
+    * 500-1000ms
+    * 1s-2.5s
+    * \>2.5s
+  * Mutation
+    * <1ms
+    * 1-10ms
+    * 10-100ms
+    * 100-500ms
+    * 500-1000ms
+    * 1s-2.5s
+    * \>2.5s
+* Data Service, Durable
+  * Mutation
+    * 1-10ms
+    * 10-100ms
+    * 100-1000ms
+    * 1-2s
+    * 2s-5s
+    * 5s-10s
+    * \>10s
+* Query Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* FTS Service
+  * Search (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
+* Analytics Service
+  * Query (all)
+    * 0-100ms
+    * 100-1000ms
+    * 1s-10s
+    * 10s-30s
+    * 30-75s
+    * \>75s
```

Comment:
If we're recording metrics on a per-request basis (which is where we've landed I think based on meeting discussions), I think would be recording metrics for the underlying KV & query requests that a transaction performs instead of anything transactions-specific.

Reply:
Transactions is a purely client-side concept. We collect metrics that are visible to the server only.